Add inital support for kfp inputPaths

ajbozarth opened the pull request:
Front-end work for https://github.com/elyra-ai/elyra/pull/2094

Add support for setting a property to a parent node's output for kfp nodes.

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

ajbozarth commented:
@akchinSTC

ajbozarth commented:
Remaining TODO: add tests for NestedEnumControl

bourdakos1 reviewed (changes requested):
couple small things, but I also have some other questions

marthacryan reviewed (commented):
Small comment but otherwise the code looks good to me - sounds like Nick has some questions though so I'll let him approve before we merge this

ajbozarth commented:
Quick status update, after a couple Webex discussions with @bourdakos1 we decided there was no reason to move any of the code onto the elyra side. The use of `inputpath` and `outputpath` format values can be generically leveraged despite those names originating from kfp, thus @bourdakos1 argued that there's no reason to push them out of pipeline-editor. Since this lines up with my original implementation I have "reverted" back to having all the UI code in this PR and will be removing the UI code from https://github.com/elyra-ai/elyra/pull/2094 these changes will be pushed as soon as I finish testing them locally.

Also included in my changes are small updates addressing review comments and moving the common code in EnumControl and NestedEnumControl into a common util file.

marthacryan commented on packages/pipeline-editor/src/PipelineEditor/index.tsx:
```diff
@@ -635,6 +635,75 @@ const PipelineEditor = forwardRef(
+      );
+
+      for (const upstreamNode of upstreamNodes) {
+        // @ts-ignore
```
Why are we adding this ignore?

ajbozarth replied:
we have to have the ignore because we're accessing `op` on the selected node, but as far as the compiler knows the node could be a super node which doesn't have the property `op`. This case can't happen though because this function is called  after a node type check to make sure it's an execution node. @bourdakos1 if there's a better way to handle this I'm open to it.

bourdakos1 replied:
we should definitely avoid the ts-ignore. There are a couple ways we can get around it. The easiest is just casting upstreamNodes to a list of execution nodes, but there is probably a better solution

bourdakos1 replied:
I think my comment disappeared...? but yea, we should avoid ts-ignore. Easiest way is to just cast upstreamNodes to a list of execution nodes, but that is kind of a bandaid and there is probably a better long term solution

ajbozarth replied:
I did a little bit of googling into the best ways to handle type assertion in typescript and the method that seem to match our use case the best is to cast `(upstreamNode as ExecutionNodeDef)?.op` inline. This cast will be `undefined` if the type doesn't match, which in turn would fail the check in the find.

ajbozarth replied:
I've pushed the above fix, feel free to review again

marthacryan reviewed (approved):
These changes look good!

bourdakos1 commented on packages/pipeline-editor/src/CustomFormControls/utils.ts:
```diff
@@ -0,0 +1,96 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
can we call this file "components" or something similar? utils isn't very descriptive

ajbozarth replied:
sure I'll change it

bourdakos1 reviewed (changes requested):
I have a couple notes, but I won't have time to flesh them out until monday

bourdakos1 commented on packages/pipeline-editor/src/PipelineEditor/index.tsx:
```diff
@@ -635,6 +636,76 @@ const PipelineEditor = forwardRef(
+        getUpstreamNodeIds(selectedNode)
+      );
+
+      for (const upstreamNode of upstreamNodes) {
+        const nodeDef = nodes.find(
+          (n) => n.op === (upstreamNode as ExecutionNodeDef)?.op
+        );
+
+        const options = [];
+        for (const prop of nodeDef?.app_data.properties?.uihints
+          ?.parameter_info ?? []) {
+          if (prop.data.format === "outputpath") {
+            options.push({
+              value: prop.parameter_ref,
+              label: prop.label.default,
+            });
+          }
+        }
+        data.push({
+          value: upstreamNode.id,
+          label: upstreamNode.app_data?.ui_data?.label,
+          options: options,
+        });
+      }
```
maybe add a couple comments describing this a bit

ptitzler commented:
Scenario: a node requires an input but it is not connected to a parent

![image](https://user-images.githubusercontent.com/13068832/134954216-0f30b95e-3073-4295-960d-837b3dd8a57a.png)

1. Could we  add a selection message to the drop down? (e.g. "Select an input source")
2. Would it be possible to omit `type:string` for those inputs that are of type `select`?

bourdakos1 commented:
> 2. Would it be possible to omit `type:string` for those inputs that are of type `select`?

@ptitzler I believe the `type:string` stuff is hard coded into the description string, which comes from the backend

ptitzler commented:
> > 1. Would it be possible to omit `type:string` for those inputs that are of type `select`?
> 
> @ptitzler I believe the `type:string` stuff is hard coded into the description string, which comes from the backend

Okay. Will take it up there.

ptitzler commented:
Scenario: a node is connected to two nodes as follows: The first node `DO THIS` was specified as an input to the last node. Subsequently the first node's label was renamed to `DO THAT`. Current behavior:

![image](https://user-images.githubusercontent.com/13068832/134964276-0c62d7d8-c1f5-4b4a-b381-bcff6d24ca99.png)

Expected behavior:
 - The GUI recognizes that the referenced node's label was changed and displays the new label as the selected label (instead of as a new sourcew)

ptitzler commented:
Do we want to be more explicit about the select option labels? Current text is `<node label/name>: <property-name> (output)`:

![image](https://user-images.githubusercontent.com/13068832/134966619-e4bfb312-f9bd-4465-96c9-bcf5858e749f.png)

It might be a tad more user friendly if we were to show something like

 ```
 node: <NAME/LABEL> output: <OUTPUT-PROPERTY>
 ```

bourdakos1 commented:
> Do we want to be more explicit about the select option labels? Current text is `<node label/name>: <property-name> (output)`:
> 
> ![image](https://user-images.githubusercontent.com/13068832/134966619-e4bfb312-f9bd-4465-96c9-bcf5858e749f.png)
> 
> It might be a tad more user friendly if we were to show something like
> 
> ```
> node: <NAME/LABEL> output: <OUTPUT-PROPERTY>
> ```

Yea, I agree. This could use a bit of UI/UX love

bourdakos1 commented on packages/pipeline-editor/src/PipelineEditor/index.tsx:
```diff
@@ -635,6 +636,76 @@ const PipelineEditor = forwardRef(
       return null;
     };
 
+    const getUpstreamNodeIds = (node: any): string[] => {
+      let upstreamNodeIds: string[] = [];
+
+      for (const link of node.inputs?.[0].links || []) {
+        upstreamNodeIds.push(link.node_id_ref);
+        const upstreamNode = controller.current.findExecutionNode(
+          link.node_id_ref
+        );
+        if (upstreamNode) {
+          upstreamNodeIds = upstreamNodeIds.concat(
+            getUpstreamNodeIds(upstreamNode)
+          );
+        }
+      }
+
+      return upstreamNodeIds;
+    };
```
I will open an issue with canvas to see if it is possible to expose [this](https://github.com/elyra-ai/canvas/blob/48781597c137c4d5fc4222e5f19c1be87afca5f0/canvas_modules/common-canvas/src/object-model/object-model.js#L1869) or if its currently useable

bourdakos1 replied:
https://github.com/elyra-ai/canvas/issues/783

bourdakos1 replied:
@ajbozarth it looks like we have direct access to the objectModel via the canvas controller. As a temporary workaround, we should be able to use:
```js
this.objectModel.getHighlightObjectIds(pipelineID, nodeIDs, "upstream");
```

We should create a wrapper for this (similar to what you have here - `findUpstreamNodes(nodeID: string)`) and put it in our `PipelineController`.

I'm not sure if it makes more sense to return a list of node IDs or to return the actual nodes? I think we map everything to a real node eventually, so it might make sense to just do it upfront in this wrapper?

ptitzler commented:
It appears that on Safari a tool tip is displayed, which helps if the selector text is only partially displayed: 

![image](https://user-images.githubusercontent.com/13068832/134991340-fea4a9d4-9e4b-486a-9955-8707e5e2f0a2.png)

However, in Chrome it is not:

![image](https://user-images.githubusercontent.com/13068832/134991130-12343e61-9b50-41a5-9a95-6069020ecbc5.png)

bourdakos1 commented:
> It appears that on Safari a tool tip is displayed, which helps if the selector text is only partially displayed:
> 
> ![image](https://user-images.githubusercontent.com/13068832/134991340-fea4a9d4-9e4b-486a-9955-8707e5e2f0a2.png)
> 
> However, in Chrome it is not:
> 
> ![image](https://user-images.githubusercontent.com/13068832/134991130-12343e61-9b50-41a5-9a95-6069020ecbc5.png)

I believe this is an existing issue for all property enums. Can you open an issue for this?

ajbozarth commented:
Pushed fixes addressing the above comments, fixing:

- @bourdakos1 code comments
- updating selected value to match changes in pipeline
- remove selected value when no longer valid
- add default text for when no value is selected
- add error visualization when NestedEnum is a required field
- Added tooltip text for all enums in all browsers

Open issues:

- currently disabling lint for one line, we should figure out how to remove this.
- an invalid value is only removed once the properties editor is opened, ie if you unlink the parent currently selected but never open the properties it will leave the (now invalid) value set and it won't show as invalid in the UI. I'm unsure if this is worth dealing with

bourdakos1 commented:
> * an invalid value is only removed once the properties editor is opened, ie if you unlink the parent currently selected but never open the properties it will leave the (now invalid) value set and it won't show as invalid in the UI. I'm unsure if this is worth dealing with

@ajbozarth is the same true for other invalid properties? For example, manually editing the pipeline JSON to make a property invalid and then open the pipeline in the UI without opening the properties (easiest one to trigger would probably be changing gpu/cpu/memory to a string instead of a number)

ajbozarth commented:
> @ajbozarth is the same true for other invalid properties? For example, manually editing the pipeline JSON to make a property invalid and then open the pipeline in the UI without opening the properties (easiest one to trigger would probably be changing gpu/cpu/memory to a string instead of a number)

That is correct, this is just one of the first instances where it can occur while editing within the pipeline editor itself. Another example would be removing a user created runtime that was set

ptitzler commented:
The "not connected to other nodes" scenario works now as expected. Thank you!

![image](https://user-images.githubusercontent.com/13068832/135511860-aca1392a-7fbf-4d6d-9eb5-a04f99f22a32.png)

Although it'd be a bit wordy, perhaps we might want to consider changing to "Select an input source" or including the phrase "from a connected node" to convey the key principle that only upstream node output can be used as input. Existing KFP user's that have gotten used to the concept of implicit connections might otherwise struggle more than necessary.

ajbozarth commented:
> perhaps we might want to consider changing to "Select an input source"

I'll update to this.

ajbozarth commented:
I've pushed an update addressing the open lint issue as well as the text update suggested by Patrick. This update does require some changes I'll push to Alans PR since I've removed `label` from the persisted "value"

ptitzler commented:
There's a problem with the following scenario. The first component produces three outputs:

![image](https://user-images.githubusercontent.com/13068832/135523013-b876a40d-700d-49f6-bca4-3a81a729c509.png)

One cannot select the desired output - the selection always reverts to the first entry:

![youdontgettochoose](https://user-images.githubusercontent.com/13068832/135523412-30b6e371-19e3-43f7-81e7-8327bcc45a15.gif)

ptitzler commented:
The node's tool tip should display user friendly information, not the 'raw' data:

![image](https://user-images.githubusercontent.com/13068832/135523915-1608663f-af09-4191-b659-2d718da5a256.png)

ptitzler commented:
> latest ...
![image](https://user-images.githubusercontent.com/13068832/135526854-d0b0a422-0e9c-4f71-ac8e-2d40bb1d2884.png)

ajbozarth commented:
> One cannot select the desired output - the selection always reverts to the first entry:

Fixed this in my last push

I also added a validator for NestedEnum, but it seems to have no added any functionality as I had hoped. 

Specifically there is still no validation before opening properties for a node. This validation occurs in the pipeline-services package and has no access to `data` to compare to to see if the current value is still valid. I am unsure how to move forward with this.

Lastly I looked into

> The node's tool tip should display user friendly information, not the 'raw' data:

This fix is also not so simple because this occurs in NodeTooltip and also doesn't have access to `data` to generate the user friendly label (since it is no longer persisted in the value)

ajbozarth commented:
Opened https://github.com/elyra-ai/elyra/issues/2184 to track the validation issue above. 

@ptitzler opened https://github.com/elyra-ai/elyra/issues/2183 to track the tool tip issue above.

With those followups open this issue should be ready for final review and merge

bourdakos1 commented on packages/pipeline-editor/src/properties-panels/NodeProperties.tsx:
```diff
@@ -30,8 +31,10 @@ interface Props {
+  upstreamNodes?: any[];
   onFileRequested?: (options: any) => any;
   onPropertiesUpdateRequested?: (options: any) => any;
+  getUpstreamNodes?: (node: any) => any;
```
```suggestion
```
